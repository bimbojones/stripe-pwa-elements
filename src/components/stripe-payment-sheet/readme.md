# stripe-payment-sheet

<!-- Auto Generated Below -->


## Properties

| Property                           | Attribute                               | Description                                                                                                                                                                                                                                                | Type                                                      | Default                          |
| ---------------------------------- | --------------------------------------- | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------- | -------------------------------- |
| `applicationName`                  | `application-name`                      | Overwrite the application name that registered For wrapper library (like Capacitor)                                                                                                                                                                        | `string`                                                  | `'stripe-pwa-elements'`          |
| `buttonLabel`                      | `button-label`                          | Submit button label By default we recommended to use these string - 'Pay' -> PaymentSheet - 'Add' -> PaymentFlow(Android) - 'Add card' -> PaymentFlow(iOS) - 'Add a card' -> PaymentFlow(iOS) These strings will translated automatically by this library. | `string`                                                  | `'Pay'`                          |
| `handleSubmit`                     | --                                      | Form submit event handler                                                                                                                                                                                                                                  | `(event: Event, props: FormSubmitEvent) => Promise<void>` | `undefined`                      |
| `intentClientSecret`               | `intent-client-secret`                  | The client secret from paymentIntent.create response                                                                                                                                                                                                       | `string`                                                  | `undefined`                      |
| `intentType`                       | `intent-type`                           | Default submit handle type. If you want to use `setupIntent`, should update this attribute.                                                                                                                                                                | `"payment" \| "setup"`                                    | `'payment'`                      |
| `publishableKey`                   | `publishable-key`                       | Your Stripe publishable API key.                                                                                                                                                                                                                           | `string`                                                  | `undefined`                      |
| `sheetTitle`                       | `sheet-title`                           | Payment sheet title By default we recommended to use these string - 'Add your payment information' -> PaymentSheet / PaymentFlow(Android) - 'Add a card' -> PaymentFlow(iOS) These strings will translated automatically by this library.                  | `string`                                                  | `'Add your payment information'` |
| `shouldUseDefaultFormSubmitAction` | `should-use-default-form-submit-action` | The component will provide a function to call the `stripe.confirmCardPayment`API. If you want to customize the behavior, should set false. And listen the 'formSubmit' event on the element                                                                | `boolean`                                                 | `true`                           |
| `showLabel`                        | `show-label`                            | Show the form label                                                                                                                                                                                                                                        | `boolean`                                                 | `false`                          |
| `showPaymentRequestButton`         | `show-payment-request-button`           | If show PaymentRequest Button, should put true                                                                                                                                                                                                             | `boolean`                                                 | `undefined`                      |
| `stripeAccount`                    | `stripe-account`                        | Optional. Making API calls for connected accounts                                                                                                                                                                                                          | `string`                                                  | `undefined`                      |
| `stripeDidLoaded`                  | --                                      | Stripe.js class loaded handler                                                                                                                                                                                                                             | `(event: StripeLoadedEvent) => Promise<void>`             | `undefined`                      |
| `zip`                              | `zip`                                   | If true, show zip code field                                                                                                                                                                                                                               | `boolean`                                                 | `true`                           |


## Events

| Event                     | Description                                   | Type                                                                                                                                                                                                                                       |
| ------------------------- | --------------------------------------------- | ------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------ |
| `defaultFormSubmitResult` | Recieve the result of defaultFormSubmit event | `CustomEvent<Error \| { paymentIntent: PaymentIntent; error?: undefined; } \| { paymentIntent?: undefined; error: StripeError; } \| { setupIntent: SetupIntent; error?: undefined; } \| { setupIntent?: undefined; error: StripeError; }>` |
| `formSubmit`              | Form submit event                             | `CustomEvent<{ stripe: Stripe; cardNumberElement: StripeCardNumberElement; cardExpiryElement: StripeCardExpiryElement; cardCVCElement: StripeCardCvcElement; intentClientSecret?: string; zipCode?: string; }>`                            |
| `stripeLoaded`            | Stripe Client loaded event                    | `CustomEvent<{ stripe: Stripe; }>`                                                                                                                                                                                                         |


## Methods

### `initStripe(publishableKey: string, options?: InitStripeOptions) => Promise<void>`

Get Stripe.js, and initialize elements

#### Returns

Type: `Promise<void>`



### `setErrorMessage(errorMessage: string) => Promise<this>`

Set error message

#### Returns

Type: `Promise<this>`



### `setPaymentRequestOption(option: PaymentRequestButtonOption) => Promise<this>`



#### Returns

Type: `Promise<this>`



### `updateProgress(progress: ProgressStatus) => Promise<this>`

Update the form submit progress

#### Returns

Type: `Promise<this>`




## Dependencies

### Used by

 - [stripe-payment-sheet](../stripe-payment-sheet-modal)

### Depends on

- [stripe-payment-request-button](../stripe-payment-request-button)

### Graph
```mermaid
graph TD;
  stripe-payment --> stripe-payment-request-button
  stripe-payment-sheet --> stripe-payment
  style stripe-payment fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
