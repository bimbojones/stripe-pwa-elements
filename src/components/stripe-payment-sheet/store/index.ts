import { loadStripe } from "@stripe/stripe-js";
import { stripeStore } from "./store";
import { PWAStripeCardElement } from "./CardElement";

export * from './store'
export * from './CardElement'
  const updateStripeAppInfo = () => {
    const stripe = stripeStore.get('stripe')
    if (!stripe) return
    const applicationName = stripeStore.get('applicationName')
    stripe.registerAppInfo({
      name: applicationName,
    });
  }
/**
 * Load StripeCardElement instance
 * @returns PWAStripeCardElement
 */
export const getAndLoadCardElement = () => {
    const elements = stripeStore.get('elements')
    const el = stripeStore.get('el')
    const element = PWAStripeCardElement.getInstance({
      el,
      elements,
    })
    return element
}
export const configureStripeJSClient = async () => {
  // If Stripe.js already loaded, do nothing
  if (!!stripeStore.get('stripe')) {
    return
  }
  stripeStore.set('loadStripeStatus', 'loading')
  const stripeAccount = stripeStore.get('stripeAccount')
  const apiKey = stripeStore.get('publishableKey')
  try {
    const stripe = await loadStripe(apiKey, {
      stripeAccount,
    })
    updateStripeAppInfo()
    stripeStore.set('stripe', stripe)
    stripeStore.set('loadStripeStatus', 'success')

    /**
     * Init Stripe Elements
     */
    const elements = stripe.elements()
    stripeStore.set('elements', elements)
    getAndLoadCardElement()
  } catch (e) {
    console.log(e)
    stripeStore.set('errorMessage', e.message)
    stripeStore.set('loadStripeStatus', 'failure')
  }

}


stripeStore.onChange('stripeAccount', configureStripeJSClient)
/**
   * Update Stripe application name
 */
stripeStore.onChange('applicationName', updateStripeAppInfo)
/**
   * Init Stripe.js
 */
stripeStore.onChange('publishableKey', configureStripeJSClient)
stripeStore.onChange('applicationName', updateStripeAppInfo)
