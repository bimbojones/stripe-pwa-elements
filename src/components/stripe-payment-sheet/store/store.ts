import { Stripe, StripeElements } from "@stripe/stripe-js";
import { createStore } from '@stencil/store'
import { ProgressStatus } from "../../../interfaces";

export type StripeStoreState = {
    /**
     * Optional. Making API calls for connected accounts
     * @info https://stripe.com/docs/connect/authentication
     */
    stripeAccount?: string;
    /**
     * Your Stripe publishable API key.
     */
    publishableKey?: string;
  
    /**
     * Status of the Stripe client initilizing process
     */
    loadStripeStatus: ProgressStatus
  
    /**
     * Stripe client class
     */
    stripe?: Stripe;
  
    /**
     * Error message
     */
    errorMessage: string;
  
    /**
     * Overwrite the application name that registered
     * For wrapper library (like Capacitor)
     */
    applicationName: string;
  
    elements?: StripeElements
    el?: HTMLStripePaymentElement
  }
  const stripeStoreInitialState: StripeStoreState = {
    loadStripeStatus: '',
    errorMessage: '',
    applicationName: 'stripe-pwa-elements',
  }

export const stripeStore = createStore<StripeStoreState>(stripeStoreInitialState)
