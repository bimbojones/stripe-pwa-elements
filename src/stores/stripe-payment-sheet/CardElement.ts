import { StripeCardCvcElement, StripeCardExpiryElement, StripeCardNumberElement, StripeElements, } from "@stripe/stripe-js";
import { i18n } from "../../utils/i18n";
import { waitForElm } from "../../utils/utils";
import { stripeStore } from "./store";

export type PWAStripeCardElementProps = {
    el: HTMLStripePaymentElement
    elements: StripeElements
  }
export  class PWAStripeCardElement {
    private el: HTMLStripePaymentElement
    private elements: StripeElements
    private static _instance: PWAStripeCardElement;
    public cardNumber!: StripeCardNumberElement;
    public cardExpiry!: StripeCardExpiryElement;
    public cardCVC!: StripeCardCvcElement;
    public static getInstance(props: PWAStripeCardElementProps): PWAStripeCardElement {
      if (!this._instance) {
        this._instance = new PWAStripeCardElement(props)
      }
      return this._instance
    }
    private constructor(props: PWAStripeCardElementProps) {
      this.el = props.el
      this.elements = props.elements
      this.init()
    }
    /**
     * Init each Card Elements
     */
    private async init() {
      const { elements, el } = this 
      const handleCardError = ({error}) => {
        if (error) {
          stripeStore.set('errorMessage', error.message)
          return
        }
        stripeStore.set('errorMessage', '')
      }
      this.cardNumber = elements.create('cardNumber', {
        placeholder: i18n.t('Card Number'),
      });
  
      const cardNumberElement: HTMLElement = await waitForElm(el, '#card-number');
  
      this.cardNumber.mount(cardNumberElement);
      this.cardNumber.on('change', handleCardError);
  
      this.cardExpiry = elements.create('cardExpiry');
      const cardExpiryElement: HTMLElement = await waitForElm(el, '#card-expiry');
  
      this.cardExpiry.mount(cardExpiryElement);
      this.cardExpiry.on('change', handleCardError);
  
      this.cardCVC = elements.create('cardCvc');
      const cardCVCElement: HTMLElement = await waitForElm(el, '#card-cvc');
  
      this.cardCVC.mount(cardCVCElement);
      this.cardCVC.on('change', handleCardError);
    }

    public unmount() {
      if (this.cardNumber) {
        this.cardNumber.unmount();
      }
  
      if (this.cardExpiry) {
        this.cardExpiry.unmount();
      }
  
      if (this.cardCVC) {
        this.cardCVC.unmount();
      }
  
    }
  }
  