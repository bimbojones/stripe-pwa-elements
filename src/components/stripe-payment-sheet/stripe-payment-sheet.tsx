import { Component, Prop, h, State, Method, EventEmitter, Event, Element } from '@stencil/core';
import { createStore } from '@stencil/store'
import { loadStripe, Stripe, StripeCardCvcElement, StripeCardExpiryElement, StripeCardNumberElement, StripeElements } from '@stripe/stripe-js';
import {checkPlatform, waitForElm} from '../../utils/utils';
import {
  StripeDidLoadedHandler,
  StripeLoadedEvent,
  FormSubmitEvent,
  FormSubmitHandler,
  ProgressStatus,
  PaymentRequestButtonOption,
  IntentType,
  DefaultFormSubmitResult,
} from '../../interfaces';
import { i18n } from '../../utils/i18n';
import { stripeStore } from '../../stores/stripe-payment-sheet/store';
import { PWAStripeCardElement } from '../../stores/stripe-payment-sheet/CardElement';
import { getAndLoadCardElement } from '../../stores/stripe-payment-sheet';

@Component({
  tag: 'stripe-payment',
  styleUrl: 'stripe-payment-sheet.scss',
  shadow: false,
})
export class StripePayment {
  @Element() el: HTMLStripePaymentElement;

  /**
   * Status of the Stripe client initilizing process
   */
  @State() loadStripeStatus: ProgressStatus = '';

  /**
   * Stripe client class
   */
  @State() stripe: Stripe;

  /**
   * Default submit handle type.
   * If you want to use `setupIntent`, should update this attribute.
   */
  @Prop() intentType: IntentType = 'payment';

  /**
   * If true, show zip code field
   */
  @Prop() zip = true;
  /**
   * Payment sheet title
   * By default we recommended to use these string
   * - 'Add your payment information' -> PaymentSheet / PaymentFlow(Android)
   * - 'Add a card' -> PaymentFlow(iOS)
   * These strings will translated automatically by this library.
   */
  @Prop() sheetTitle = 'Add your payment information';

  /**
   * Submit button label
   * By default we recommended to use these string
   * - 'Pay' -> PaymentSheet
   * - 'Add' -> PaymentFlow(Android)
   * - 'Add card' -> PaymentFlow(iOS)
   * - 'Add a card' -> PaymentFlow(iOS)
   * These strings will translated automatically by this library.
   *
   */
  @Prop() buttonLabel = 'Pay';

  /**
   * Get Stripe.js, and initialize elements
   * @param publishableKey
   * @param options
   * @example
   * ```
   * const stripeElement = document.createElement('stripe-card-element');
   * customElements
   *  .whenDefined('stripe-card-element')
   *  .then(() => {
   *    tripeElement.initStripe('pk_test_XXXXXXXXX')
   *  })
   * ```
   */
  @Method()
  public async initStripe(
    publishableKey: string,
    options: {
      stripeAccount?: string;
    } = undefined,
  ) {
    const stripeAccount = options?.stripeAccount

    stripeStore.set('el', this.el)
    stripeStore.set('stripeAccount', stripeAccount)
    stripeStore.set('applicationName', this.applicationName)
    stripeStore.set('publishableKey' , publishableKey)
    stripeStore.onChange('loadStripeStatus', async newState => {
      if (newState !== 'success') {return;}
      await this.initElement()
      this.stripeLoadedEventHandler()
    })
  }

  /**
   * The progress status of the checkout process
   */
  @State() progress: ProgressStatus = '';

  /**
   * Update the form submit progress
   * @param progress
   * @returns
   * @example
   * ```
   * const stripeElement = document.createElement('stripe-card-element');
   * customElements
   *  .whenDefined('stripe-card-element')
   *  .then(() => {
   *    // You must set the attributes to stop running default form submit action when you want to listen the 'formSubmit' event.
   *    stripeElement.setAttribute('should-use-default-form-submit-action', false)
   *    stripeElement.addEventListener('formSubmit', async props => {
   *      const {
   *        detail: { stripe, cardNumber, event },
   *      } = props;
   *      const result = await stripe.createPaymentMethod({
   *        type: 'card',
   *        card: cardNumber,
   *      });
   *      console.log(result);
   *      stripeElement.updateProgress('success')
   *    });
   * })
   */
  @Method()
  public async updateProgress(progress: ProgressStatus) {
    this.progress = progress;
    return this;
  }

  /**
   * Error message
   */
  @State() errorMessage = '';

  /**
   * zip code
   */
  @State() zipCode = '';

  /**
   * Set error message
   * @param errorMessage string
   * @returns
   * @example
   * ```
   * const stripeElement = document.createElement('stripe-card-element');
   * customElements
   *  .whenDefined('stripe-card-element')
   *  .then(() => {
   *    // You must set the attributes to stop running default form submit action when you want to listen the 'formSubmit' event.
   *    stripeElement.setAttribute('should-use-default-form-submit-action', false)
   *    stripeElement.addEventListener('formSubmit', async props => {
   *      try {
   *        throw new Error('debug')
   *      } catch (e) {
   *        stripeElement.setErrorMessage(`Error: ${e.message}`)
   *        stripeElement.updateProgress('failure')
   *      }
   *   });
   * })
   */
  @Method()
  public async setErrorMessage(errorMessage: string) {
    this.errorMessage = errorMessage;
    return this;
  }

  /**
   * Your Stripe publishable API key.
   */
  @Prop() publishableKey: string;

  /**
   * Optional. Making API calls for connected accounts
   * @info https://stripe.com/docs/connect/authentication
   */
  @Prop() stripeAccount: string;

  /**
   * Overwrite the application name that registered
   * For wrapper library (like Capacitor)
   */
  @Prop() applicationName = 'stripe-pwa-elements';

  /**
   * Show the form label
   */
  @Prop() showLabel = false;

  /**
   * The client secret from paymentIntent.create response
   *
   * @example
   * ```
   * const stripeElement = document.createElement('stripe-card-element');
   * customElements
   *  .whenDefined('stripe-card-element')
   *  .then(() => {
   *     stripeElement.setAttribute('intent-client-secret', 'dummy')
   *   })
   * ```
   *
   * @example
   * ```
   * <stripe-card-element intent-client-secret="dummy" />
   * ```
   */
  @Prop() intentClientSecret?: string;

  /**
   * The component will provide a function to call the `stripe.confirmCardPayment`API.
   * If you want to customize the behavior, should set false.
   * And listen the 'formSubmit' event on the element
   */
  @Prop() shouldUseDefaultFormSubmitAction = true;

  /**
   * If show PaymentRequest Button, should put true
   */
  @Prop()
  showPaymentRequestButton: boolean;

  @State() paymentRequestOption?: PaymentRequestButtonOption;
  /**
   * @param option
   * @private
   */
  @Method()
  public async setPaymentRequestOption(option: PaymentRequestButtonOption) {
    this.paymentRequestOption = option;
    this.createPaymentRequestButton();
    return this;
  }

  /**
   * Form submit event handler
   */
  @Prop({
    mutable: true,
  })
  handleSubmit: FormSubmitHandler;

  /**
   * Stripe.js class loaded handler
   */
  @Prop({
    mutable: true,
  })
  stripeDidLoaded?: StripeDidLoadedHandler;

  /**
   * Stripe Client loaded event
   * @example
   * ```
   * const stripeElement = document.createElement('stripe-card-element');
   * customElements
   *  .whenDefined('stripe-card-element')
   *  .then(() => {
   *     stripeElement
   *      .addEventListener('stripeLoaded', async ({ detail: {stripe} }) => {
   *       stripe
   *         .createSource({
   *           type: 'ideal',
   *           amount: 1099,
   *           currency: 'eur',
   *           owner: {
   *             name: 'Jenny Rosen',
   *           },
   *           redirect: {
   *             return_url: 'https://shop.example.com/crtA6B28E1',
   *           },
   *         })
   *         .then(function(result) {
   *           // Handle result.error or result.source
   *         });
   *       });
   *   })
   * ```
   */
  @Event() stripeLoaded: EventEmitter<StripeLoadedEvent>;
  private stripeLoadedEventHandler() {
    const event: StripeLoadedEvent = {
      stripe: this.stripe,
    };

    if (this.stripeDidLoaded) {
      this.stripeDidLoaded(event);
    }

    this.stripeLoaded.emit(event);
  }

  /**
   * Form submit event
   * @example
   * ```
   * const stripeElement = document.createElement('stripe-card-element');
   * customElements
   *  .whenDefined('stripe-card-element')
   *  .then(() => {
   *     stripeElement
   *       .addEventListener('formSubmit', async props => {
   *         const {
   *           detail: { stripe, cardNumber, event },
   *         } = props;
   *         const result = await stripe.createPaymentMethod({
   *           type: 'card',
   *           card: cardNumber,
   *         });
   *         console.log(result);
   *       })
   *   })
   */
  @Event() formSubmit: EventEmitter<FormSubmitEvent>;
  private async formSubmitEventHandler() {
    const { cardCVC, cardExpiry, cardNumber, stripe } = this;

    this.formSubmit.emit({
      cardCVCElement: cardCVC,
      cardExpiryElement: cardExpiry,
      cardNumberElement: cardNumber,
      stripe,
    });
  }

  /**
   * Recieve the result of defaultFormSubmit event
   * @example
   * ```
   * const stripeElement = document.createElement('stripe-card-element');
   * customElements
   *  .whenDefined('stripe-card-element')
   *  .then(() => {
   *     stripeElement.addEventListener('defaultFormSubmitResult', async ({detail}) => {
   *       if (detail instanceof Error) {
   *         console.error(detail)
   *       } else {
   *         console.log(detail)
   *       }
   *     })
   *   })
   */
  @Event() defaultFormSubmitResult: EventEmitter<DefaultFormSubmitResult>;
  private async defaultFormSubmitResultHandler(result: DefaultFormSubmitResult) {
    this.defaultFormSubmitResult.emit(result);
  }

  private cardNumber!: StripeCardNumberElement;
  private cardExpiry!: StripeCardExpiryElement;
  private cardCVC!: StripeCardCvcElement;

  componentWillUpdate() {
    if (!this.publishableKey) {
      return;
    }

    if (['success', 'loading'].includes(this.loadStripeStatus)) {
      return;
    }

    this.initStripe(this.publishableKey, {
      stripeAccount: this.stripeAccount,
    });
    this.createPaymentRequestButton();
  }

  /**
   * Default form submit action (just call a confirmCardPayment).
   * If you don't want use it, please set `should-use-default-form-submit-action="false"`
   * @param event
   * @param param1
   */
  private async defaultFormSubmitAction(event: Event, { stripe, cardNumberElement, intentClientSecret }: FormSubmitEvent) {
    event.preventDefault();
    try {
      const { intentType } = this;
      const result = await (() => {
        if (intentType === 'payment') {
          return stripe.confirmCardPayment(intentClientSecret, {
            payment_method: {
              card: cardNumberElement,
            },
          });
        }

        return stripe.confirmCardSetup(intentClientSecret, {
          payment_method: {
            card: cardNumberElement,
          },
        });
      })();

      this.defaultFormSubmitResultHandler(result);
    } catch (e) {
      this.defaultFormSubmitResultHandler(e);
      throw e;
    }
  }

  constructor() {
    if (this.publishableKey) {
      this.initStripe(this.publishableKey, {
        stripeAccount: this.stripeAccount,
      });
    } else {
      this.loadStripeStatus = 'failure';
    }
  }

  /**
   * Initialize Component using Stripe Element
   */
  private async initElement() {
    document.getElementById('stripe-card-element').addEventListener('submit', async e => {
      const elements = getAndLoadCardElement();
      const { cardCVC, cardExpiry, cardNumber } = elements
      const stripe = stripeStore.get('stripe')
      const { intentClientSecret } = this

      const submitEventProps: FormSubmitEvent = {
        cardCVCElement: cardCVC,
        cardExpiryElement: cardExpiry,
        cardNumberElement: cardNumber,
        stripe,
        intentClientSecret,
        zipCode: this.zipCode,
      };

      this.progress = 'loading';
      try {
        if (this.handleSubmit) {
          await this.handleSubmit(e, submitEventProps);
        } else if (this.shouldUseDefaultFormSubmitAction === true && intentClientSecret) {
          await this.defaultFormSubmitAction(e, submitEventProps);
        } else {
          e.preventDefault();
        }

        await this.formSubmitEventHandler();
        if (this.handleSubmit || this.shouldUseDefaultFormSubmitAction === true) {
          this.progress = 'success';
        }
      } catch (e) {
        this.errorMessage = e.message;
        this.progress = 'failure';
      }
    });
  }
  componentDidLoad() {
    this.el.classList.add(checkPlatform());
  }

  disconnectedCallback() {
    const el = stripeStore.get('el')
    const elements = stripeStore.get('elements')
    const pwaStripeCardElement = PWAStripeCardElement.getInstance({
      el,
      elements,
    })

    pwaStripeCardElement.unmount()
  }

  /**
   * Create payment request button
   * It's just proxy of stripe-payment-request-button
   */
  private createPaymentRequestButton() {
    const { showPaymentRequestButton, paymentRequestOption } = this;

    if (!showPaymentRequestButton || !paymentRequestOption) {
      return null;
    }

    if (!document) {
      return null;
    }

    const targetElement = document.getElementById('stripe-payment-request-button');
    const stripePaymentRequestElement = document.createElement('stripe-payment-request-button');

    targetElement.appendChild(stripePaymentRequestElement);

    const { paymentRequestPaymentMethodHandler, paymentRequestShippingOptionChangeHandler, paymentRequestShippingAddressChangeHandler } = paymentRequestOption;

    customElements.whenDefined('stripe-payment-request-button').then(() => {
      stripePaymentRequestElement.setPaymentRequestOption(paymentRequestOption);

      if (paymentRequestPaymentMethodHandler) {
        stripePaymentRequestElement.setPaymentMethodEventHandler(paymentRequestPaymentMethodHandler);
      }

      if (paymentRequestShippingOptionChangeHandler) {
        stripePaymentRequestElement.setPaymentRequestShippingOptionEventHandler(paymentRequestShippingOptionChangeHandler);
      }

      if (paymentRequestShippingAddressChangeHandler) {
        stripePaymentRequestElement.setPaymentRequestShippingAddressEventHandler(paymentRequestShippingAddressChangeHandler);
      }

      return stripePaymentRequestElement.initStripe(this.publishableKey);
    });
  }

  render() {
    const { errorMessage } = this;

    if (this.loadStripeStatus === 'failure') {
      return <p>{i18n.t('Failed to load Stripe')}</p>;
    }

    const disabled = this.progress === 'loading';

    return (
      <div class="stripe-payment-sheet-wrap">
        <form id="stripe-card-element">
          <div class="stripe-heading">{i18n.t(this.sheetTitle)}</div>
          <div id="stripe-payment-request-button" />
          <div>
            <div class="stripe-section-title">{i18n.t('Card information')}</div>
          </div>
          <div class="payment-info card visible">
            <fieldset class="stripe-input-box">
              <div>
                <label>
                  {this.showLabel ? <lenged>{i18n.t('Card Number')}</lenged> : null}
                  <div id="card-number" />
                </label>
              </div>
              <div class="stripe-input-column" style={{ display: 'flex' }}>
                <label style={{ width: '50%' }}>
                  {this.showLabel ? <lenged>{i18n.t('MM / YY')}</lenged> : null}
                  <div id="card-expiry" />
                </label>
                <label style={{ width: '50%' }}>
                  {this.showLabel ? <lenged>{i18n.t('CVC')}</lenged> : null}
                  <div id="card-cvc" />
                </label>
              </div>
              <div id="card-errors" class="element-errors">
                {errorMessage}
              </div>
            </fieldset>
          </div>
          {this.zip ? (
            <div style={{ marginTop: '1.5rem' }}>
              <div class="stripe-section-title">{i18n.t('Country or region')}</div>
            </div>
          ) : null}
          {this.zip ? (
            <div class="payment-info card visible">
              <fieldset class="stripe-input-box">
                <div>
                  <label>
                    {this.showLabel ? <lenged>{i18n.t('Postal Code')}</lenged> : null}
                    <input
                      id="zip"
                      name="zip"
                      type="text"
                      inputmode="numeric"
                      class="stripe-input-box StripeElement"
                      style={{ width: '100%' }}
                      placeholder={i18n.t('Postal Code')}
                      value={this.zipCode}
                      onInput={e => {
                        this.zipCode = (e.target as any).value;
                      }}
                    />
                  </label>
                </div>
              </fieldset>
            </div>
          ) : null}
          <div style={{ marginTop: '32px' }}>
            <button type="submit" disabled={disabled}>
              {this.progress === 'loading' ? i18n.t('Loading') : i18n.t(this.buttonLabel)}
            </button>
          </div>
        </form>
      </div>
    );
  }
}
