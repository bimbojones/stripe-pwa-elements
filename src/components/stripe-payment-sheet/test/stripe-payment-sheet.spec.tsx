import { newSpecPage } from '@stencil/core/testing';
import { StripePayment } from '../stripe-payment-sheet';
import { stripeStore } from '../store';

describe('stripe-payment', () => {
  beforeEach(() => {
    stripeStore.dispose();
  });
  it('renders', async () => {
    const page = await newSpecPage({
      components: [StripePayment],
      html: `<stripe-payment publishable-key='pk_test_xxx'></stripe-payment>`,
    });

    expect(page.root).toEqualHtml(`
 <stripe-payment class="undefined"  publishable-key='pk_test_xxx'>
      <div class="stripe-payment-sheet-wrap">
        <form id="stripe-card-element">
          <div class="stripe-heading">
            Add your payment information
          </div>
          <div id="stripe-payment-request-button"></div>
          <div>
            <div class="stripe-section-title">
              Card information
            </div>
          </div>
          <div class="card payment-info visible">
            <fieldset class="stripe-input-box">
              <div>
                <label>
                  <div id="card-number"></div>
                </label>
              </div>
              <div class="stripe-input-column" style="display: flex;">
                <label style="width: 50%;">
                  <div id="card-expiry"></div>
                </label>
                <label style="width: 50%;">
                  <div id="card-cvc"></div>
                </label>
              </div>
              <div class="element-errors" id="card-errors"></div>
            </fieldset>
          </div>
          <div style="margin-top: 1.5rem;">
            <div class="stripe-section-title">
              Country or region
            </div>
          </div>
          <div class="card payment-info visible">
            <fieldset class="stripe-input-box">
              <div>
                <label>
                  <input class="StripeElement stripe-input-box" id="zip" inputmode="numeric" name="zip" placeholder="Postal Code" type="text" value="" style="width: 100%;">
                </label>
              </div>
            </fieldset>
          </div>
          <div style="margin-top: 32px;">
            <button type="submit">
              Pay
            </button>
          </div>
        </form>
      </div>
    </stripe-payment>
    `);
  });
  it('render [No zip code]', async () => {
    const page = await newSpecPage({
      components: [StripePayment],
      html: `<stripe-payment zip="false"  publishable-key='pk_test_xxx'></stripe-payment>`,
    });

    expect(page.root).toEqualHtml(`
 <stripe-payment class="undefined" zip="false"  publishable-key='pk_test_xxx'>
      <div class="stripe-payment-sheet-wrap">
        <form id="stripe-card-element">
          <div class="stripe-heading">
            Add your payment information
          </div>
          <div id="stripe-payment-request-button"></div>
          <div>
            <div class="stripe-section-title">
              Card information
            </div>
          </div>
          <div class="card payment-info visible">
            <fieldset class="stripe-input-box">
              <div>
                <label>
                  <div id="card-number"></div>
                </label>
              </div>
              <div class="stripe-input-column" style="display: flex;">
                <label style="width: 50%;">
                  <div id="card-expiry"></div>
                </label>
                <label style="width: 50%;">
                  <div id="card-cvc"></div>
                </label>
              </div>
              <div class="element-errors" id="card-errors"></div>
            </fieldset>
          </div>
          <div style="margin-top: 32px;">
            <button type="submit">
              Pay
            </button>
          </div>
        </form>
      </div>
    </stripe-payment>
    `);
  });
});
