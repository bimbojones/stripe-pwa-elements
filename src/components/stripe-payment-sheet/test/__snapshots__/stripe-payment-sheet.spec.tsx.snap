// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stripe-payment rendering test api key and zip code props 1`] = `
<stripe-payment class="undefined" publishable-key="pk_test_xxx" zip="false">
  <div class="stripe-payment-sheet-wrap">
    <form id="stripe-card-element">
      <div class="stripe-heading">
        Add your payment information
      </div>
      <div id="stripe-payment-request-button"></div>
      <div>
        <div class="stripe-section-title">
          Card information
        </div>
      </div>
      <div class="card payment-info visible">
        <fieldset class="stripe-input-box">
          <div>
            <label>
              <div id="card-number"></div>
            </label>
          </div>
          <div class="stripe-input-column" style="display: flex;">
            <label style="width: 50%;">
              <div id="card-expiry"></div>
            </label>
            <label style="width: 50%;">
              <div id="card-cvc"></div>
            </label>
          </div>
          <div class="stripe-element-errors" id="card-errors"></div>
        </fieldset>
      </div>
      <div style="margin-top: 32px;">
        <button type="submit">
          Pay
        </button>
      </div>
    </form>
  </div>
</stripe-payment>
`;

exports[`stripe-payment rendering test should load stripe after setting the publishable-key (snapshot) 1`] = `
<stripe-payment '="" class="undefined" publishable-key="yyyy" zip="false">
  <div class="stripe-payment-sheet-wrap">
    <form id="stripe-card-element">
      <div class="stripe-heading">
        Add your payment information
      </div>
      <div id="stripe-payment-request-button"></div>
      <div>
        <div class="stripe-section-title">
          Card information
        </div>
      </div>
      <div class="card payment-info visible">
        <fieldset class="stripe-input-box">
          <div>
            <label>
              <div id="card-number"></div>
            </label>
          </div>
          <div class="stripe-input-column" style="display: flex;">
            <label style="width: 50%;">
              <div id="card-expiry"></div>
            </label>
            <label style="width: 50%;">
              <div id="card-cvc"></div>
            </label>
          </div>
          <div class="stripe-element-errors" id="card-errors"></div>
        </fieldset>
      </div>
      <div style="margin-top: 32px;">
        <button type="submit">
          Pay
        </button>
      </div>
    </form>
  </div>
</stripe-payment>
`;

exports[`stripe-payment rendering test with the api key 1`] = `
<stripe-payment class="undefined" publishable-key="pk_test_xxx">
  <div class="stripe-payment-sheet-wrap">
    <form id="stripe-card-element">
      <div class="stripe-heading">
        Add your payment information
      </div>
      <div id="stripe-payment-request-button"></div>
      <div>
        <div class="stripe-section-title">
          Card information
        </div>
      </div>
      <div class="card payment-info visible">
        <fieldset class="stripe-input-box">
          <div>
            <label>
              <div id="card-number"></div>
            </label>
          </div>
          <div class="stripe-input-column" style="display: flex;">
            <label style="width: 50%;">
              <div id="card-expiry"></div>
            </label>
            <label style="width: 50%;">
              <div id="card-cvc"></div>
            </label>
          </div>
          <div class="stripe-element-errors" id="card-errors"></div>
        </fieldset>
      </div>
      <div style="margin-top: 1.5rem;">
        <div class="stripe-section-title">
          Country or region
        </div>
      </div>
      <div class="card payment-info visible">
        <fieldset class="stripe-input-box">
          <div>
            <label>
              <input class="StripeElement stripe-input-box" id="zip" inputmode="numeric" name="zip" placeholder="Postal Code" type="text" value="" style="width: 100%;">
            </label>
          </div>
        </fieldset>
      </div>
      <div style="margin-top: 32px;">
        <button type="submit">
          Pay
        </button>
      </div>
    </form>
  </div>
</stripe-payment>
`;
